File: smtp/typescript/examples/oauth2-gmail.ts
/**
 * Gmail OAuth2 Authentication Example
 *
 * This example demonstrates how to send emails through Gmail using
 * OAuth2 authentication (XOAUTH2 SASL mechanism).
 *
 * Prerequisites:
 * 1. Create a project in Google Cloud Console
 * 2. Enable Gmail API
 * 3. Create OAuth 2.0 credentials (OAuth client ID)
 * 4. Get a refresh token using the OAuth 2.0 Playground or your own flow
 */

import {
  smtpClient,
  EmailBuilder,
  TlsMode,
  AuthMethod,
  createOAuth2Token,
  createOAuth2Authenticator,
  OAuth2Provider,
  OAuth2Token,
} from '@integrations/smtp';

/**
 * Gmail OAuth2 configuration
 * Replace these values with your actual credentials
 */
const GMAIL_CONFIG = {
  clientId: 'your-client-id.apps.googleusercontent.com',
  clientSecret: 'your-client-secret',
  refreshToken: 'your-refresh-token',
  userEmail: 'your-email@gmail.com',
};

/**
 * Custom OAuth2 provider that refreshes access tokens
 */
class GmailOAuth2Provider implements OAuth2Provider {
  private clientId: string;
  private clientSecret: string;
  private refreshToken: string;
  private userEmail: string;
  private cachedToken: OAuth2Token | null = null;
  private tokenExpiry: number = 0;

  constructor(config: typeof GMAIL_CONFIG) {
    this.clientId = config.clientId;
    this.clientSecret = config.clientSecret;
    this.refreshToken = config.refreshToken;
    this.userEmail = config.userEmail;
  }

  async getToken(): Promise<OAuth2Token> {
    // Return cached token if still valid
    const now = Date.now();
    if (this.cachedToken && now < this.tokenExpiry) {
      console.log('Using cached OAuth2 token');
      return this.cachedToken;
    }

    console.log('Refreshing OAuth2 token...');

    // Request a new access token from Google
    const response = await fetch('https://oauth2.googleapis.com/token', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: new URLSearchParams({
        client_id: this.clientId,
        client_secret: this.clientSecret,
        refresh_token: this.refreshToken,
        grant_type: 'refresh_token',
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Failed to refresh OAuth2 token: ${error}`);
    }

    const data = await response.json();

    // Create and cache the new token
    this.cachedToken = createOAuth2Token({
      user: this.userEmail,
      accessToken: data.access_token,
    });

    // Set token expiry (subtract 5 minutes for safety)
    const expiresIn = data.expires_in || 3600;
    this.tokenExpiry = now + (expiresIn - 300) * 1000;

    console.log('OAuth2 token refreshed successfully');
    return this.cachedToken;
  }
}

async function main() {
  // Create OAuth2 provider
  const oauth2Provider = new GmailOAuth2Provider(GMAIL_CONFIG);

  // Create OAuth2 authenticator
  const authenticator = createOAuth2Authenticator(oauth2Provider);

  // Create SMTP client configured for Gmail
  const client = smtpClient()
    .host('smtp.gmail.com')
    .port(587) // Gmail SMTP port with STARTTLS
    .tlsMode(TlsMode.StartTls)
    .authMethod(AuthMethod.XOAuth2)
    .authenticator(authenticator)
    .connectTimeout(15000)
    .commandTimeout(30000)
    .build();

  try {
    // Build the email
    const email = new EmailBuilder()
      .from(GMAIL_CONFIG.userEmail)
      .to('recipient@example.com')
      .subject('Test Email via Gmail OAuth2')
      .text('This email was sent using Gmail SMTP with OAuth2 authentication.')
      .html(`
<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
</head>
<body>
  <h2>Gmail OAuth2 Test</h2>
  <p>This email was sent using Gmail SMTP with OAuth2 authentication.</p>
  <p>Benefits of OAuth2:</p>
  <ul>
    <li>More secure than password authentication</li>
    <li>Tokens can be revoked without changing password</li>
    <li>Granular permissions</li>
    <li>No need to enable "Less secure app access"</li>
  </ul>
</body>
</html>
      `)
      .build();

    console.log('Sending email via Gmail with OAuth2...');

    // Send the email
    const result = await client.send(email);

    console.log('Email sent successfully!');
    console.log('Message ID:', result.messageId);

  } catch (error) {
    console.error('Failed to send email via Gmail:', error);
    throw error;
  } finally {
    await client.close();
  }
}

/**
 * Example: Send multiple emails using connection pooling
 */
async function sendBatchWithGmail() {
  const oauth2Provider = new GmailOAuth2Provider(GMAIL_CONFIG);
  const authenticator = createOAuth2Authenticator(oauth2Provider);

  // Enable connection pooling for better performance
  const client = smtpClient()
    .host('smtp.gmail.com')
    .port(587)
    .tlsMode(TlsMode.StartTls)
    .authMethod(AuthMethod.XOAuth2)
    .authenticator(authenticator)
    // Configure connection pool
    .poolConfig({
      maxConnections: 3, // Gmail may rate limit, so use fewer connections
      minIdleConnections: 1,
      maxIdleTime: 60000,
      acquireTimeout: 10000,
    })
    // Configure rate limiting to respect Gmail's sending limits
    .rateLimitConfig({
      maxPerSecond: 2, // Send max 2 emails per second
      maxPerMinute: 50, // Gmail free tier: ~100/minute, be conservative
      maxPerHour: 500,
      onLimitBehavior: 'throttle' as const,
    })
    .build();

  try {
    const recipients = [
      'user1@example.com',
      'user2@example.com',
      'user3@example.com',
      'user4@example.com',
      'user5@example.com',
    ];

    const emails = recipients.map(recipient =>
      new EmailBuilder()
        .from(GMAIL_CONFIG.userEmail)
        .to(recipient)
        .subject('Newsletter - December 2025')
        .text('Check out our latest newsletter!')
        .build()
    );

    console.log(`Sending ${emails.length} emails via Gmail...`);

    // Send batch
    const result = await client.sendBatch(emails);

    console.log('Batch send complete!');
    console.log('Successful:', result.successful.length);
    console.log('Failed:', result.failed.length);

    if (result.failed.length > 0) {
      console.error('Failed emails:');
      for (const failed of result.failed) {
        console.error(`  - ${failed.email.to[0].address}: ${failed.error.message}`);
      }
    }

  } finally {
    await client.close();
  }
}

/**
 * Important notes for using Gmail:
 *
 * 1. Sending Limits:
 *    - Free Gmail: 500 emails/day, ~100/minute
 *    - Google Workspace: 2000 emails/day, ~200/minute
 *
 * 2. OAuth2 Setup:
 *    - Use Google Cloud Console to create OAuth credentials
 *    - Add https://mail.google.com/ to scopes
 *    - Use OAuth 2.0 Playground to get refresh token
 *
 * 3. Security:
 *    - Store credentials in environment variables or secure vault
 *    - Never commit OAuth credentials to version control
 *    - Rotate tokens regularly
 *
 * 4. Rate Limiting:
 *    - Gmail enforces rate limits
 *    - Use the rate limiter to avoid hitting limits
 *    - Monitor for 421 and 450 response codes
 *
 * 5. Best Practices:
 *    - Enable connection pooling for batch sends
 *    - Implement retry logic with exponential backoff
 *    - Monitor metrics for failures
 *    - Use SPF, DKIM, and DMARC for better deliverability
 */

// Run the example
if (require.main === module) {
  main().catch((error) => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

export { main, sendBatchWithGmail, GmailOAuth2Provider };

File: smtp/rust/examples/oauth2_gmail.rs
//! Gmail OAuth2 Authentication Example
//!
//! This example demonstrates how to:
//! - Configure XOAUTH2 authentication for Gmail
//! - Obtain OAuth2 access tokens
//! - Connect to Gmail's SMTP server
//! - Send emails using OAuth2 credentials

use smtp_integration::{
    SmtpClient, SmtpConfig, Email, EmailAddress, AuthMethod, SmtpError,
};

#[tokio::main]
async fn main() -> Result<(), SmtpError> {
    // Gmail SMTP server settings
    const GMAIL_SMTP_HOST: &str = "smtp.gmail.com";
    const GMAIL_SMTP_PORT: u16 = 587;

    // OAuth2 credentials
    // In a real application, you would:
    // 1. Register your app in Google Cloud Console
    // 2. Enable Gmail API
    // 3. Create OAuth 2.0 credentials
    // 4. Implement OAuth2 flow to get access token
    // 5. Refresh tokens when they expire

    // For this example, we'll show how to use an access token
    // Note: This is a placeholder - you need to implement the OAuth2 flow
    let oauth2_access_token = get_oauth2_access_token().await?;
    let user_email = "your-email@gmail.com";

    // Configure SMTP client with OAuth2
    let config = SmtpConfig::builder()
        .host(GMAIL_SMTP_HOST)
        .port(GMAIL_SMTP_PORT)
        .auth_method(AuthMethod::OAuth2 {
            user: user_email.to_string(),
            access_token: oauth2_access_token,
        })
        .use_tls(true)
        .use_starttls(true)
        .timeout_seconds(30)
        .build()?;

    println!("Creating Gmail SMTP client with OAuth2...");
    let client = SmtpClient::new(config)?;

    // Build the email
    let email = Email::builder()
        .from(EmailAddress::new(user_email, Some("Your Name")))
        .to(EmailAddress::new("recipient@example.com", Some("Recipient")))
        .subject("Test Email via Gmail OAuth2")
        .text_body("This email was sent using Gmail's SMTP server with OAuth2 authentication.")
        .html_body(r#"
<!DOCTYPE html>
<html>
<head>
    <style>
        body { font-family: Arial, sans-serif; }
        .success { color: #4CAF50; font-weight: bold; }
    </style>
</head>
<body>
    <h2>OAuth2 Authentication Success!</h2>
    <p class="success">This email was sent using Gmail's SMTP server with OAuth2 authentication.</p>
    <p>This demonstrates secure, token-based authentication without storing passwords.</p>
</body>
</html>
"#)
        .build()?;

    println!("Sending email via Gmail...");
    println!("  From: {}", email.from.email);
    println!("  To: {}", email.to.iter().map(|a| a.email.as_str()).collect::<Vec<_>>().join(", "));
    println!("  Subject: {}", email.subject);
    println!("  Auth: OAuth2");

    // Send the email
    match client.send(&email).await {
        Ok(response) => {
            println!("Email sent successfully via Gmail!");
            println!("  Message ID: {:?}", response.message_id);
            println!("  Status: {}", response.status);
        }
        Err(e) => {
            eprintln!("Failed to send email: {}", e);
            return Err(e);
        }
    }

    Ok(())
}

/// Get OAuth2 access token for Gmail
///
/// In a real application, you would implement the full OAuth2 flow:
/// 1. Redirect user to Google's authorization endpoint
/// 2. User grants permission
/// 3. Exchange authorization code for access token and refresh token
/// 4. Store refresh token securely
/// 5. Use refresh token to get new access tokens when they expire
///
/// This is a simplified example showing the structure.
async fn get_oauth2_access_token() -> Result<String, SmtpError> {
    // Example OAuth2 flow (simplified)
    // You would typically use a library like `oauth2` or `yup-oauth2`

    println!("\n=== OAuth2 Setup Instructions ===");
    println!("1. Go to Google Cloud Console: https://console.cloud.google.com/");
    println!("2. Create a new project or select an existing one");
    println!("3. Enable Gmail API");
    println!("4. Create OAuth 2.0 credentials (Desktop app or Web app)");
    println!("5. Download the credentials JSON file");
    println!("\n6. Implement OAuth2 flow:");
    println!("   - Authorization URL: https://accounts.google.com/o/oauth2/v2/auth");
    println!("   - Token URL: https://oauth2.googleapis.com/token");
    println!("   - Scopes: https://mail.google.com/");
    println!("\n7. Example using `yup-oauth2` crate:");
    println!("   ```rust");
    println!("   use yup_oauth2::{{InstalledFlowAuthenticator, InstalledFlowReturnMethod}};");
    println!("   ");
    println!("   let secret = yup_oauth2::read_application_secret(\"credentials.json\").await?;");
    println!("   let auth = InstalledFlowAuthenticator::builder(");
    println!("       secret,");
    println!("       InstalledFlowReturnMethod::HTTPRedirect,");
    println!("   )");
    println!("   .build()");
    println!("   .await?;");
    println!("   ");
    println!("   let scopes = &[\"https://mail.google.com/\"];");
    println!("   let token = auth.token(scopes).await?;");
    println!("   ```");
    println!("================================\n");

    // In a real implementation, you would:
    // 1. Load credentials from file or environment
    // 2. Check if we have a valid cached token
    // 3. If not, perform OAuth2 flow
    // 4. Return the access token

    // For this example, we'll read from an environment variable
    std::env::var("GMAIL_OAUTH2_ACCESS_TOKEN")
        .map_err(|_| SmtpError::AuthenticationFailed(
            "GMAIL_OAUTH2_ACCESS_TOKEN environment variable not set.\n\
             Please obtain an OAuth2 access token and set it in the environment.\n\
             See the instructions above for details.".to_string()
        ))
}

// Advanced example: Using yup-oauth2 for token management
// Uncomment and add `yup-oauth2` to your Cargo.toml to use this

/*
use yup_oauth2::{InstalledFlowAuthenticator, InstalledFlowReturnMethod};

async fn get_oauth2_access_token_with_library() -> Result<String, SmtpError> {
    // Read OAuth2 credentials from file
    let secret = yup_oauth2::read_application_secret("credentials.json")
        .await
        .map_err(|e| SmtpError::ConfigurationError(format!("Failed to read credentials: {}", e)))?;

    // Create authenticator
    let auth = InstalledFlowAuthenticator::builder(
        secret,
        InstalledFlowReturnMethod::HTTPRedirect,
    )
    .persist_tokens_to_disk("tokens.json")
    .build()
    .await
    .map_err(|e| SmtpError::ConfigurationError(format!("Failed to create authenticator: {}", e)))?;

    // Get token (this will open a browser for first-time authorization)
    let scopes = &["https://mail.google.com/"];
    let token = auth
        .token(scopes)
        .await
        .map_err(|e| SmtpError::AuthenticationFailed(format!("Failed to get token: {}", e)))?;

    Ok(token.token().unwrap().to_string())
}
*/
